### PR TITLE
Issue6592 - di header file created even if errors occur

### PR DESCRIPTION
**Pull request by cristiancreteanu**
Moved header file generation at the end to make sure the header file is generate only if no errors occur during compilation.

### CODE REVIEW

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
Thank you for the pull request, but there are some guidelines when contributing to DMD:

 - When fixing an issue you should put "Fix Issue $Issue_number ..." in the commit message so that the bot automatically links your PR to the bugzilla report and automatically closes the issue when the PR is merged.

- Every PR should contain a test file that tests the changed behavior. This is to ensure that future patches do not change the added/established behavior. Take a look in the tests directory from the root of the repo.

Lastly, the patch is a breaking change. Header generation needs to be done before semantic analysis, because semantic usually modifies the AST, therefore influencing the output of the generated header files. More over, it is a performance regression in the situations where you don't care about actually compiling the code, but you want to simply generate header files.

I think that the right fix is to generate the header file, but after semantic, simply check if any errors occurred. If that is the case, then simply delete the file.

**Comment by WalterBright** [comment]
I agree with @RazvanN7

File: src/dmd/dmodule.d [hunk 1]
```diff
@@ -695,7 +695,7 @@ extern (C++) final class Module : Package
             }
             else
                 fprintf(stderr, "Specify path to file '%s' with -I switch\n", srcfile.toChars());
-            fatal();
+            // fatal();
```
Comment (RazvanN7) [inline src/dmd/dmodule.d]: wju is this commented? Maybe it should calll `removeHdrFilesAndFail`?

Comment (RazvanN7) [reply src/dmd/dmodule.d]: Oh, I see. If it fails, the code in mars.d will catch that.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
Congratulations on your first dmd PR! Auto-merge-squash toggled on.

**Pull request merged**